File: 240628/2차원 최대 증가 수열/longest-increasing-sequence-2d.cpp
#include <iostream>
#include <algorithm>
#include <climits>
using namespace std;
#define MAX_NM 50

int n, m, ans;
int grid[MAX_NM][MAX_NM], dp[MAX_NM][MAX_NM];

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
            dp[i][j] = INT_MIN;
        }
    }

    for(int i=0; i<n; i++) {
        dp[i][0] = 1;
        dp[0][i] = 1;
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            for(int r=0; r<i; r++) {
                for(int c=0; c<j; c++) {
                    if(grid[r][c] < grid[i][j]) {
                        dp[i][j] = max(dp[i][j], dp[r][c]+1);
                    }
                }
            }
            ans = max(ans, dp[i][j]);
        }
    }

    cout << ans;
    return 0;
}

File: 250218/2차원 최대 증가 수열/longest-increasing-sequence-2d.py
import sys

INT_MIN = -sys.maxsize

n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

dp = [[INT_MIN]*m for _ in range(n)]

dp[0][0] = 1

for x in range(1, n):
    for y in range(1, m):
        for i in range(x):
            for j in range(y):
                if grid[i][j] is not INT_MIN and grid[i][j] < grid[x][y]:
                    dp[x][y] = max(dp[x][y], dp[i][j]+1)


ans = INT_MIN
for i in range(n):
    for j in range(m):
        ans = max(ans, dp[i][j])

print(ans)
